File: src/cleansia_android/core/src/main/java/cz/cleansia/core/ui/components/CleansiaChip.kt
package cz.cleansia.core.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.selection.toggleable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

/**
 * A selectable pill. The one chip the customer app draws, wherever it offers a short closed set of
 * choices — cancellation reasons, review tags.
 *
 * **It takes a resolved [label], never a string resource id.** `:core` cannot reach an app's `R`, and
 * the two apps' copy is deliberately their own. Mirrors the iOS `CleansiaChip`.
 *
 * **Selection policy belongs to the caller, not here.** Cancellation reasons are radio-style with
 * tap-to-clear; review tags are multi-select with a cap. Both are one line at the call site, and
 * folding either into this component would make it wrong for the other — the original private version
 * hardwired single-select and could not be shared because of it.
 */
@Composable
fun CleansiaChip(
    label: String,
    isSelected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
    val border = if (isSelected) MaterialTheme.colorScheme.primary
                 else MaterialTheme.colorScheme.outlineVariant
    val bg = if (isSelected) MaterialTheme.colorScheme.primary.copy(alpha = 0.12f)
             else MaterialTheme.colorScheme.surface
    val labelColor = if (isSelected) MaterialTheme.colorScheme.primary
                     else MaterialTheme.colorScheme.onSurface
    Box(
        modifier = modifier
            .clip(RoundedCornerShape(999.dp))
            .background(bg)
            .border(
                width = if (isSelected) 1.5.dp else 1.dp,
                color = border,
                shape = RoundedCornerShape(999.dp),
            )
            // toggleable, NOT clickable(role = Role.Checkbox). The role alone tells a screen reader
            // what KIND of control this is; only `value` tells it the current state. With clickable
            // TalkBack says "checkbox" and never "checked", so a blind customer hears the same thing
            // for a tag they picked and one they did not — which is the whole thing this needs to say.
            .toggleable(
                value = isSelected,
                enabled = enabled,
                role = Role.Checkbox,
                onValueChange = { onClick() },
            )
            .padding(horizontal = 14.dp, vertical = 8.dp),
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium.copy(
                fontWeight = if (isSelected) FontWeight.SemiBold else FontWeight.Normal,
            ),
            color = labelColor,
        )
    }
}

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/Components/CleansiaChip.swift
import SwiftUI

/// A selectable pill. The one chip the customer app draws wherever it offers a short closed set of
/// choices — cancellation reasons, review tags. Compose `CleansiaChip` parity, geometry for geometry.
///
/// **Selection policy belongs to the caller.** Cancellation reasons are radio-style with tap-to-clear;
/// review tags are multi-select under a cap. Both are one line at the call site, and folding either in
/// here would make the component wrong for the other — which is exactly why the original private
/// version could not be shared.
///
/// Lay a row of these out with ``ChipFlow``, not a fixed n-per-row grid: the labels are localized into
/// five languages and a Czech or Ukrainian label is routinely half again the width of its English
/// original.
public struct CleansiaChip: View {
    private let label: String
    private let isSelected: Bool
    private let enabled: Bool
    private let action: () -> Void

    public init(
        label: String,
        isSelected: Bool,
        enabled: Bool = true,
        action: @escaping () -> Void
    ) {
        self.label = label
        self.isSelected = isSelected
        self.enabled = enabled
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            Text(label)
                .font(CleansiaTypography.bodyMedium)
                .foregroundColor(isSelected ? CleansiaColors.primary : CleansiaColors.onSurface)
                // Truncate rather than wrap — same rule as the buttons, same as the Android twin.
                .lineLimit(1)
                .padding(.horizontal, Spacing.s)
                .padding(.vertical, Spacing.xs)
                .background(
                    isSelected ? CleansiaColors.primary.opacity(0.12) : CleansiaColors.surface,
                    in: Capsule()
                )
                .overlay(
                    Capsule().stroke(
                        isSelected ? CleansiaColors.primary : CleansiaColors.outlineVariant,
                        lineWidth: isSelected ? 1.5 : 1
                    )
                )
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        // isSelected has to be announced, not just tappable: a multi-select chip row read out as a
        // list of buttons leaves a VoiceOver user no way to tell which tags they already picked.
        .accessibilityAddTraits(isSelected ? [.isButton, .isSelected] : .isButton)
    }
}
